Automatically determine top level domains

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds some code to automatically determine the list of top-level domains (e.g. .gov, .com, .edu, etc.) and pass it to the [`gather` script](https://github.com/18F/domain-scan#gathering-hostnames) from [18F/domain-scan](https://github.com/cisagov/18F/domain-scan).

## 💭 Motivation and context ##

Resolves #52.

Several top-level domains were added recently via cisagov/scan-target-data#45 and cisagov/scan-target-data#48, and I realized this morning that we have not been scanning some of them because their top-level domain was not included here:
https://github.com/cisagov/gatherer/blob/e17fdca7e5e3cf0dd56295e498480717aa59a97b/gather-domains.sh#L75-L99

Automatically extracting the list of domains is a much more maintainable and much less error-prone solution than continuing to add them manually.

## 🧪 Testing ##

I built a new Docker container and ran the new code in our production CyHy environment.  It functions as expected.

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

jsf9k committed:
Automatically determine the list of top-level domains to use

Several non-.gov top-level domains have been added recently via
cisagov/scan-target-data#45 and cisagov/scan-target-data#48.
Automatically extracting the list of domains is a much more
maintainable and much less error-prine solution than continuing to add
manually them.

jsf9k commented:
The Travis build is not kicking off because [Travis is no longer free](https://docs.travis-ci.com/user/billing-overview/#usage---credits), and we have exhausted our 10000 free credits.

mcdonnnj commented on gather-domains.sh:
```diff
@@ -82,12 +82,21 @@ sed -i '/[^,]*,[^,]*,Library of Congress,/d;/[^,]*,[^,]*,Government Publishing O
-# --suffix argument because of the domains included in
-# include/current-federal-non-dotgov.csv
-###
+# Note that we have to include .gov as well as.edu, .com, and other
```
```suggestion
# Note that we have to include .gov as well as .edu, .com, and other
```

dav3r reviewed (approved):
Nice bugfix and improvement! 👍

mcdonnnj commented on gather-domains.sh:
```diff
@@ -82,12 +82,21 @@ sed -i '/[^,]*,[^,]*,Library of Congress,/d;/[^,]*,[^,]*,Government Publishing O
+###
+TOP_LEVEL_DOMAINS=$(cut --delimiter=, --fields=1 $OUTPUT_DIR/current-federal_modified.csv | \
+                        tail --lines=+2 | \
+                        sed "s/^.*\(\.[^\.]*\)$/\1/" | \
```
Thoughts on using `awk` instead of a `sed` regex?

```suggestion
                        awk -F"." '{print $NF;}' | \
```

jsf9k replied:
I tend to use `sed` just because I know regex better than I know `awk` syntax.  

1. I also need the leading `.`, so the `awk` code should be `awk -F"." '{print "."$NF;}'`.
2. I can also skip the first line in the same `awk` code:  `awk -F"." '{if (NR > 1) print "."$NF;}'`.  This lets me get rid of the `tail` command.

What do you think of that?

mcdonnnj replied:
Even better! I forgot about the leading `.` so apologies there. I like doing the `tail` work here as well. I tend toward regex as a default too, I just like avoiding it for simpler tasks like this.

jsf9k replied:
See commit 2de3935.

mcdonnnj replied:
Thanks for refining my suggestion!

mcdonnnj commented on gather-domains.sh:
```diff
@@ -82,12 +82,21 @@ sed -i '/[^,]*,[^,]*,Library of Congress,/d;/[^,]*,[^,]*,Government Publishing O
+TOP_LEVEL_DOMAINS=$(cut --delimiter=, --fields=1 $OUTPUT_DIR/current-federal_modified.csv | \
+                        tail --lines=+2 | \
+                        sed "s/^.*\(\.[^\.]*\)$/\1/" | \
+                        tr "[:upper:]" "[:lower:]" | \
+                        sort | \
+                        uniq | \
```
Just a thought.

```suggestion
                        sort --ignore-case --unique | \
                        tr "[:upper:]" "[:lower:]" | \
```

mcdonnnj reviewed (approved):
Thanks for improving this and updating it to require no manual maintenance. 💪💪💪 I had one small typo to fix and a couple thoughts.

jsf9k committed:
Fix a typo

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Combine sort and uniq pipeline stages into a single stage

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Combine tail and sed pipeline stages into a single stage using awk

The awk code is also simpler than the sed regex.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Bump base image up to Python 3.9

I re-tested that the resulting image functions as expected.

mcdonnnj reviewed (approved):
Even better than the first time ✔

jsf9k committed:
Bumping version from 1.4.0 to 1.4.1

jsf9k committed:
Temporarily turn off beautysh for a piece of bash code

beautysh doesn't seem to be able to correctly handle this statement
and mangles the rest of the file if allowed to parse it.
